Add `publish` command support

rmarticaba opened the pull request:
Add support for publishing messages using pub / sub paradigm instead of storing them into database with rpush. It is created through configuration and tests are added to validate message reception and contents.
Revert username to it's previous `default` value.

rmarticaba commented:
I created this new Pull Request to add another feature we needed in my current project. We wanted to use the redis logging through Redis pub/sub protocol instead of storing the messages in database. This way we avoided polling to get new messages and could show them directly. 
I also set again username's default value as `default` through config so it keeps backwards compatibility.
Please let me know what do you think of these changes.

rmarticaba commented on appender/src/main/java/com/vlkan/log4j2/redis/appender/RedisAppender.java:
```diff
@@ -337,7 +349,7 @@ public static class Builder implements org.apache.logging.log4j.core.util.Builde
 
         @PluginBuilderAttribute
-        private String username = null;
+        private String username = "default";
```
I think this helps keeping backwards compatibility

rmarticaba commented on appender/src/test/java/com/vlkan/log4j2/redis/appender/JedisPubSubTestImpl.java:
This is a helper class the keep count of the messages on subscribe

vy replied:
Could you make the following changes, please?

* Rename the class to `JedisSubscriber`
* Make the class package-private and `final`
* Remove `onPong()` and `onSubscribe()` methods
* Make `messageCount()` and `nextMessage()` package-private
* Mark all methods `synchronized`
* Remove `super.*()` calls

rmarticaba replied:
Done

rmarticaba commented on appender/src/test/java/com/vlkan/log4j2/redis/appender/RedisServerExtension.java:
```diff
@@ -37,7 +37,7 @@ class RedisServerExtension implements BeforeEachCallback, AfterEachCallback {
                     .port(port)
                     .bind("0.0.0.0")
-                    .setting("user " + username + " on -debug +@all ~* >" + password)
+                    .setting("user " + username + " on -debug +@all ~* &* >" + password)
```
Allowing all pub/sub channels

vy replied:
I am pretty sure next week I won't recall what these cryptic symbols mean:
```suggestion
                    // Add user, grant access rights, and set password: https://redis.io/docs/latest/operate/oss_and_stack/management/security/acl
                    .setting("user " + username + " on -debug +@all allkeys allchannels >" + password)
```

vy commented on appender/src/test/java/com/vlkan/log4j2/redis/appender/RedisAppenderPubSubTest.java:
```diff
@@ -0,0 +1,196 @@
+
+    private final int redisPort = NetworkUtils.findUnusedPort(redisHost);
+
+    private final String redisPassword = String.format("%s-RedisPubPassword-%s:%d", CLASS_NAME, redisHost, redisPort);
+
+    private final String redisUsername = String.format("%s-RedisPubUsername-%s", CLASS_NAME, redisHost);
```
```suggestion
    private final String redisPassword = String.format("%s-RedisPassword-%s:%d", CLASS_NAME, redisHost, redisPort);

    private final String redisUsername = String.format("%s-RedisUsername-%s", CLASS_NAME, redisHost);
```

vy commented on appender/src/test/java/com/vlkan/log4j2/redis/appender/RedisAppenderPubSubTest.java:
```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2017-2023 Volkan Yazıcı
```
```suggestion
 * Copyright 2017-2024 Volkan Yazıcı
```

vy commented on appender/src/test/java/com/vlkan/log4j2/redis/appender/RedisAppenderPubSubTest.java:
```diff
@@ -0,0 +1,196 @@
+                .addAttribute("password", redisPassword)
+                .addAttribute("key", redisKey)
+                .addAttribute("ignoreExceptions", false)
+                .addAttribute("usePubSubParadigm", true)
```
```suggestion
                .addAttribute("command", "publish")
```

vy commented on appender/src/test/java/com/vlkan/log4j2/redis/appender/RedisAppenderPubSubTest.java:
```diff
@@ -0,0 +1,196 @@
+                .getLogger(RedisAppenderPubSubTest.class);
+
+        final JedisPubSubTestImpl pubsub = new JedisPubSubTestImpl();
+        executor.execute(() -> redisClientExtension.getClient().subscribe(pubsub, redisKey));
```
Why do we need to invoke `subscribe()` in the background? Why doesn't calling `subscribe()` in the current thread work for the purposes of the test?

rmarticaba replied:
Because it is a blocking call and the test does not progress, we need a thread to be subscribed for the messages to be sent, but not the current one.

vy reviewed (changes requested):
@rmarticaba, using `publish` was something I had in mind for quite a while too. Thanks so much for the contribution! :100:

rmarticaba commented on README.md:
```diff
@@ -102,8 +102,9 @@ Note that `sentinelNodes` and `sentinelMaster` have priority over `host` and
-| `username` | String | `null` | Redis username |
+| `username` | String | `default` | Redis username |
 | `password` | String | `null` | Redis password |
+| `usePubSubParadigm` | boolean | `false` | Publish messages to a channel identified by `key` instead of using rpush. |
```
Thanks for the feedback. I see the improvements and they make totally sense.

rmarticaba commented:
> @rmarticaba, using `publish` was something I had in mind for quite a while too. Thanks so much for the contribution! 💯

Thanks, I think I addressed all your comments and pushed the changes. It's a nice tool and I'm glad to have been able to contribute to it.

vy commented:
@rmarticaba, I have converted `JedisTestSubscriber` to a JUnit extension. This not only safely attaches a background thread to the test<sup>1</sup>, it also doesn't leak the thread<sup>2</sup>. All in all, thanks so much for the contribution. I will cut the `0.16.0` release right away.

<sup>1</sup> Since we register the thread shutdown to a `AfterEachCallback`.
<sup>2</sup>  You were shutting down the executor, which signals an interrupt to worker threads. This is not enough to wake `subscribe()` up from its sleep:`unsubscribe()` needs to be called too.
